File: LR0/src/Types/FractionType.java
package Types;

import java.io.Serializable;

public class FractionType implements Serializable {
    private Integer integerPartValue;
    private Integer numeratorValue;
    private Integer denominatorValue;

    private Boolean isNegative;

    public FractionType(Boolean isNegative, Integer integerPartValue, Integer numeratorValue, Integer denominatorValue) throws InstantiationException {
        if (integerPartValue < 0 || numeratorValue < 0 || denominatorValue < 1) {
            throw new InstantiationException("Cannot be instantiated from this values!");
        }
        this.isNegative = isNegative;
        this.integerPartValue = integerPartValue;
        this.numeratorValue = numeratorValue;
        this.denominatorValue = denominatorValue;

        simplify();
    }

    private void simplify() {
        int gcd = gcd(numeratorValue, denominatorValue);
        numeratorValue /= gcd;
        denominatorValue /= gcd;

        if (numeratorValue >= denominatorValue) {
            integerPartValue += numeratorValue / denominatorValue;
            numeratorValue %= denominatorValue;
        }
    }

    public Integer getIntegerPartValue() {
        return this.integerPartValue;
    }

    public Integer getNumeratorValue() {
        return this.numeratorValue;
    }

    public Integer getDenominatorValue() {
        return this.denominatorValue;
    }

    public Boolean getIsNegative() {
        return this.isNegative;
    }

    public void setIntegerPartValue(Integer integerPartValue) throws InstantiationException {
        if (integerPartValue < 0) {
            throw new InstantiationException("Cannot be instantiated from this values!");
        }
        this.integerPartValue = integerPartValue;
        simplify();
    }

    public void setNumeratorValue(Integer numeratorValue) throws InstantiationException {
        if (numeratorValue < 0) {
            throw new InstantiationException("Cannot be instantiated from this values!");
        }
        this.numeratorValue = numeratorValue;
        simplify();
    }

    public void setDenominatorValue(Integer denominatorValue) throws InstantiationException {
        if (denominatorValue < 1) {
            throw new InstantiationException("Cannot be instantiated from this values!");
        }
        this.denominatorValue = denominatorValue;
        simplify();
    }

    public void setIsNegative(Boolean isNegative) {
        this.isNegative = isNegative;
    }

    public static Integer gcd(Integer a, Integer b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public static Integer lcm(Integer a, Integer b) {
        return (a * b) / gcd(a, b);
    }

    @Override
    public String toString() {
        return (this.isNegative ? "-" : "") + this.integerPartValue + "." + this.numeratorValue + "/" + this.denominatorValue;
    }

}

File: LR1/src/main/scala/types/FractionType.scala
package types

object FractionType {
  def gcd(a: Integer, b: Integer): Integer = {
    if (b == 0) a else gcd(b, a % b)
  }

  def lcm(a: Integer, b: Integer): Integer = (a * b) / gcd(a, b)
}

class FractionType @throws[InstantiationException]
(private var isNegative: Boolean, private var integerPartValue: Integer, private var numeratorValue: Integer, private var denominatorValue: Integer) extends Serializable {
  if (integerPartValue < 0 || numeratorValue < 0 || denominatorValue < 1) throw new InstantiationException("Cannot be instantiated from this values!")
  simplify()

  private def simplify(): Unit = {
    val gcd = FractionType.gcd(numeratorValue, denominatorValue)
    numeratorValue /= gcd
    denominatorValue /= gcd
    if (numeratorValue >= denominatorValue) {
      integerPartValue += numeratorValue / denominatorValue
      numeratorValue %= denominatorValue
    }
  }

  def getIntegerPartValue: Integer = this.integerPartValue

  def getNumeratorValue: Integer = this.numeratorValue

  def getDenominatorValue: Integer = this.denominatorValue

  def getIsNegative: Boolean = this.isNegative

  @throws[InstantiationException]
  def setIntegerPartValue(integerPartValue: Integer): Unit = {
    if (integerPartValue < 0) throw new InstantiationException("Cannot be instantiated from this values!")
    this.integerPartValue = integerPartValue
    simplify()
  }

  @throws[InstantiationException]
  def setNumeratorValue(numeratorValue: Integer): Unit = {
    if (numeratorValue < 0) throw new InstantiationException("Cannot be instantiated from this values!")
    this.numeratorValue = numeratorValue
    simplify()
  }

  @throws[InstantiationException]
  def setDenominatorValue(denominatorValue: Integer): Unit = {
    if (denominatorValue < 1) throw new InstantiationException("Cannot be instantiated from this values!")
    this.denominatorValue = denominatorValue
    simplify()
  }

  def setIsNegative(isNegative: Boolean): Unit = {
    this.isNegative = isNegative
  }

  override def toString: String = (if (this.isNegative) "-"
  else "") + this.integerPartValue + "." + this.numeratorValue + "/" + this.denominatorValue
}